src/ctapipe/calib/camera/extractor.py

```diff
@@ -0,0 +1,250 @@
+"""
+Extraction algorithms to compute the statistics from a chunk of images
+"""
+
+__all__ = [
+    "StatisticsExtractor",
+    "PlainExtractor",
+    "SigmaClippingExtractor",
+]
+
+from abc import abstractmethod
+
+import numpy as np
+from astropy.stats import sigma_clipped_stats
+
+from ctapipe.containers import StatisticsContainer
+from ctapipe.core import TelescopeComponent
+from ctapipe.core.traits import (
+    CaselessStrEnum,
+    Int,
+    List,
+)
+
+
+class StatisticsExtractor(TelescopeComponent):
+    """
+    Base component to handle the extraction of the statistics from a dl1 table
+    containing charges, peak times and/or charge variances (images).
+    """
+
+    chunk_size = Int(
+        2500,
+        help="Size of the chunk used for the calculation of the statistical values",
+    ).tag(config=True)
+
+    def __call__(
+        self,
+        dl1_table,
+        masked_pixels_of_sample=None,
+        chunk_shift=None,
+        col_name="image",
+    ) -> list:
+        """
+        Prepare the extraction chunks and call the relevant function of the particular extractor
+        to extract the statistical values.
+
+        Parameters
+        ----------
+        dl1_table : astropy.table.Table
+            DL1 table with images of shape (n_images, n_channels, n_pix)
+            and timestamps of shape (n_images, ) stored in an astropy Table
+        masked_pixels_of_sample : ndarray
+            boolean array of masked pixels of shape (n_pix, ) that are not available for processing
+        chunk_shift : int
+            number of samples to shift the extraction chunk
```

Comment:
From the description here, I didn't understand that chunk_shift shift by which the window is advanced and that `chunk_shift < chunk_size` implies overlapping chunks.

Maybe the description here could be clearer.

Reply:
I also think this should be a traitlet, so people can configure this through the config system

Reply:
The idea of making it not a trailet was that we want to pass twice over the data with only one extractor instance. Maybe this pseudo code helps (justify that to make `chunk_shift` a trailet of the component using the `StatisticsExtractor` API):

```python
        ...
        # Get the extractor
        extractor = self.stats_extractor[self.stats_extractor_type.tel[tel_id]]

        # First pass through the whole provided dl1 data
        stats_list_firstpass = extractor(dl1_table=dl1_table[tel_id], col_name=col_name)

        for chunk_nr, stats in enumerate(stats_list_firstpass):
            ...
            # Detect faulty chunks by calculating the fraction of faulty pixels over the camera and checking if the threshold is exceeded.
            outliers_to_faultless_pixels_ratio = detect_outliers(...)
            # Check if chunk is faulty 
            if outliers_to_faultless_pixels_ratio > self.faulty_pixels_threshold:
                   # Get the ranges for slicing the region of troubles
                   slice_start, slice_stop = self._get_slice_range(...)
                   # Slice the dl1 table according to the previously caluclated start and stop.
                   dl1_table_sliced = dl1_table[tel_id][slice_start:slice_stop]
                   # Run the stats extractor on the sliced dl1 table with a chunk_shift
                   # to remove the period of trouble (carflashes etc.) as effectively as possible.
                   stats_list_secondpass = extractor(
                        dl1_table=dl1_table_sliced, chunk_shift=self.chunk_shift
                    )
                    # Do something
                    do_somthing(stats_list_secondpass)

```